mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -265,101 +293,145 @@ impl Watcher {
-
-    let mut swap_fut = Box::pin(
-        async move {
-            let mut events;
-            loop {
-                let res = running_swap.handle_command(command).await.expect("!handle_command");
-                events = res.1;
-                for event in events {
-                    status.status(&[&"swap", &("uuid", uuid_str.as_str())], &event.status_str());
-                    running_swap.apply_event(event);
-                }
-                match res.0 {
-                    Some(c) => {
-                        command = c;
-                    },
-                    None => {
-                        break;
-                    },
+        let refund_fut = watcher_ctx
+            .taker_coin
+            .send_watcher_refunds_taker_payment_preimage(&watcher_ctx.data.taker_refunds_payment);
+        let transaction = match refund_fut.compat().await {
+            Ok(t) => t,
+            Err(err) => {
+                if let Some(tx) = err.get_tx() {
+                    broadcast_p2p_tx_msg(
+                        &watcher_ctx.ctx,
+                        tx_helper_topic(watcher_ctx.taker_coin.ticker()),
+                        &tx,
+                        &None,
+                    );
                 }
-            }
+
+                return Self::change_state(Stopped::from_reason(StopReason::TakerPaymentRefundFailed(
+                    ERRL!("{}", err.get_plain_text_format()).into(),
+                )));
+            },
+        };
+
+        broadcast_p2p_tx_msg(
+            &watcher_ctx.ctx,
+            tx_helper_topic(watcher_ctx.taker_coin.ticker()),
+            &transaction,
+            &None,
+        );
+
+        let wait_fut = watcher_ctx.taker_coin.wait_for_confirmations(
+            &transaction.tx_hex(),
+            1,
+            false,
+            watcher_ctx.data.taker_payment_lock + WAIT_FOR_TAKER_REFUND + 3600,
+            WAIT_CONFIRM_INTERVAL,
+        );
+        if let Err(err) = wait_fut.compat().await {
+            return Self::change_state(Stopped::from_reason(StopReason::TakerPaymentRefundFailed(
+                ERRL!("{}", err).into(),
+            )));
         }
-        .fuse(),
-    );
-    let do_nothing = (); // to fix https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit
-    select! {
-        _swap = swap_fut => do_nothing, // swap finished normally
-    };
+
+        let tx_hash = transaction.tx_hash();
+        info!("Taker refund tx hash {:02x}", tx_hash);
+        Self::change_state(Stopped::from_reason(StopReason::TakerPaymentRefunded))
+    }
+}
+
+#[async_trait]
+impl LastState for Stopped {
+    type Ctx = WatcherContext;
+    type Result = ();
+    async fn on_changed(self: Box<Self>, watcher_ctx: &mut Self::Ctx) -> Self::Result {
+        let swap_ctx = SwapsContext::from_ctx(&watcher_ctx.ctx).unwrap();
+        swap_ctx
+            .taker_swap_watchers
+            .lock()
+            .remove(watcher_ctx.data.taker_fee_hash.clone());
```

Comment:
`SwapWatcherLock` removes corresponding `taker_fee_hash` from `taker_swap_watchers` container once the state machine is finished. Do we need to explicitly do the same on `Stopped` state?

Reply:
Hmm no we don't, will remove it.